core/makeswift/components/product-detail/client.tsx

```diff
@@ -0,0 +1,84 @@
+'use client';
+
+import React, {
+  type ComponentPropsWithoutRef,
+  createContext,
+  forwardRef,
+  type PropsWithChildren,
+  type ReactNode,
+  type Ref,
+  useContext,
+  useMemo,
+} from 'react';
+
+import { ProductDetail } from '@/vibes/soul/sections/product-detail';
+
+type VibesProductDetailProps = ComponentPropsWithoutRef<typeof ProductDetail>;
+
+type Props = VibesProductDetailProps & {
+  product: VibesProductDetailProps['product'] & {
+    description: string;
+    plainTextDescription: string;
+  };
+};
+
+const PropsContext = createContext<Props | null>(null);
+
+export const PropsContextProvider = ({ value, children }: PropsWithChildren<{ value: Props }>) => (
+  <PropsContext.Provider value={value}>{children}</PropsContext.Provider>
+);
+
+export const DescriptionType = {
+  PlainText: 'PlainText',
+  RichText: 'RichText',
+  Custom: 'Custom',
+} as const;
+
+type DescriptionType = (typeof DescriptionType)[keyof typeof DescriptionType];
+
+export const MakeswiftProductDetail = forwardRef(
+  (
+    {
+      descriptionType,
+      slot,
+    }: {
+      descriptionType: DescriptionType;
+      slot: ReactNode;
+    },
+    ref: Ref<HTMLDivElement>,
+  ) => {
+    const context = useContext(PropsContext);
+    const description = useMemo(() => {
+      const product = context?.product;
+      if (product == null) return null;
+
+      switch (descriptionType) {
+        case DescriptionType.PlainText:
+          return product.plainTextDescription;
+        case DescriptionType.RichText:
+          return (
+            <div className="prose" dangerouslySetInnerHTML={{ __html: product.description }} />
+          );
+        case DescriptionType.Custom:
+          return slot;
+      }
+    }, [descriptionType, context?.product, slot]);
+
+    if (context == null) {
+      console.error('No context provided for MakeswiftProductDetail');
+      return <p ref={ref}>There was an error rendering the product detail.</p>;
+    }
```

Comment:
I'd remove this check, personally. You are already handling the null context/product gracefully, and even those we are registering these components as private, I've been structuring them in a way where you can use them as regular editable client components should decide to make them public.

Reply:
I'd love to as well. I need to do this because I didn't want to provide the default props for the `useContext`, since we have to define this:
```
type Props<F extends Field> = {
  breadcrumbs?: Breadcrumb[];
  product: ProductDetailProduct;
  action: ProductDetailFormAction<F>;
  fields: F[];
};
```

Or do you think it's still better to setup an initial data there? @agurtovoy

Reply:
Hmm, I see your dilemma now. I actually can't think of a better way to handle this, I guess we just need to be content that some embedded components can never become freestanding.

Reply:
Sounds good.